src/index.ts

```diff
@@ -0,0 +1,765 @@
+   * @param {boolean} [node.isWrapping] if true, when asking for the next child at the end or start of the node, the will "wrap around" and return the first/last (when asking for the last/first)
+   * @param {string} [node.orientation] can be "vertical" or "horizontal". is used in conjuction when handling direction of key press, to determine which child is "next"
+   * @param {boolean} [node.isIndexAlign] if a node is index aligned, its descendents should jump to nodes based on index instead of activeChild
+   * @param {function} [node.onLeave] if a node has an `onLeave` function, it will be run when a move event leaves this node
+   * @param {function} [node.onEnter] if a node has an `onEnter` function, it will be run when a move event enters this node
+   */
+  registerNode (nodeId, node : any = {}) {
+    if (!node.id) {
+      node.id = nodeId
+    }
+
+    // if this is the very first node, set it as root and return...
+    if (Object.keys(this.tree).length <= 0) {
+      this.rootNodeId = nodeId
+      this.tree[nodeId] = node
+      this.nodePathList.push(nodeId)
+      return this
+    }
+
+    // if this node DOESNT have a parent assume its parent is root
+    if (node.parent == null && nodeId !== this.rootNodeId) {
+      node.parent = this.rootNodeId
+    }
+
+    // if this node is the first child of its parent, we need to set its parent's `activeChild`
+    // to it so that the parent always has an `activeChild` value
+    // we can tell if its parent has any children by checking the nodePathList for
+    // entries containing '<parent>.children'
+    const parentsChildPaths = this.nodePathList.find(path => path.includes(node.parent + '.children'))
+    if (parentsChildPaths == null) {
+      const parentPath = this.getPathForNodeId(node.parent)
+      Set(this.tree, parentPath + '.activeChild', nodeId)
+    }
+
+    // if no `index` set, calculate it
+    if (!node.index) {
+      let parentNode = this.getNode(node.parent)
+      if (parentNode) {
+        let parentsChildren = this.getNode(node.parent).children
+        if (!parentsChildren) {
+          node.index = 0
+        } else {
+          node.index = (Object.keys(parentsChildren).length)
+        }
+      }
+    }
+
+    // add the node into the tree
+    // path is the node's parent plus 'children' plus itself
+    let path = this.nodePathList.find(path => path.endsWith(node.parent)) + '.children.' + nodeId
+    Set(this.tree, path, node)
+    this.nodePathList.push(path)
+
+    // if the node is focusable, we want to add its path to our focusableNodePathList
+    if (isFocusable(node)) {
+      this.focusableNodePathList.push(path)
+    }
+
+    return this
+  }
+
+  /**
+   * maintained for legacy API reasons
+   */
+  register(nodeId, node : any = {}) {
+    return this.registerNode(nodeId, node)
+  }
+
+  /**
+   * unregister a node from the navigation tree
+   * @param {string} nodeId
+   */
+  unregisterNode (nodeId) {
```

Comment:
Should unregistering a node also unregister any overrides that have this node as their target?

Reply:
Yep, I agree it should. Will wire up this behaviour and add some tests.

Reply:
Also added to this logic a little to unregister an `override` if its `target` OR `id` matches the ID of the node that was unregistered.